airbyte-cdk/bulk/toolkits/load-db/src/main/kotlin/io/airbyte/cdk/load/orchestration/db/legacy_typing_deduping/TableNameMapFactory.kt

```diff
@@ -32,23 +38,185 @@ class TableCatalogFactory(
+            } else {
+                processedRawTableNames.add(originalRawTableName)
+                currentRawProcessedName = originalRawTableName
+            }
+
+            if (originalFinalTableName in processedFinalTableNames) {
+                currentFinalProcessedName = TableName(originalFinalTableName.namespace, newName)
+                processedFinalTableNames.add(currentFinalProcessedName)
+            } else {
+                processedFinalTableNames.add(originalFinalTableName)
+                currentFinalProcessedName = originalFinalTableName
+            }
+
+            // Create column name mapping with collision handling
+            val columnNameMapping = createColumnNameMapping(stream)
+
+            result[stream] =
                 TableNameInfo(
                     TableNames(
-                        rawTableName = rawTableNameGenerator.getTableName(stream.descriptor),
-                        finalTableName = finalTableNameGenerator.getTableName(stream.descriptor),
+                        rawTableName = currentRawProcessedName,
+                        finalTableName = currentFinalProcessedName,
                     ),
-                    ColumnNameMapping(
-                        // TODO handle collisions in column names
-                        stream.schema.asColumns().mapValues { (columnName, _) ->
-                            finalTableColumnNameGenerator.getColumnName(columnName).displayName
-                        }
-                    )
+                    columnNameMapping
+                )
+        }
+
+        return TableCatalog(result)
+    }
+
+    /**
+     * Creates column name mapping with handling for potential collisions using incremental
+     * numbering, with advanced resolution for truncation cases.
+     */
+    private fun createColumnNameMapping(stream: DestinationStream): ColumnNameMapping {
+        val processedColumnNames = mutableSetOf<String>()
+        val columnMappings = mutableMapOf<String, String>()
+        // Map to track original column names by their truncated versions
+        val originalColumnNameMap = mutableMapOf<String, String>()
+
+        stream.schema.asColumns().forEach { (columnName, _) ->
+            val processedColumnName =
+                finalTableColumnNameGenerator.getColumnName(columnName).canonicalName
+
+            // Store mapping between processed name and original name
+            originalColumnNameMap[processedColumnName] = columnName
+
+            // Get a unique column name by adding incremental numbers if necessary
+            val finalColumnName =
+                resolveColumnNameCollision(
+                    processedColumnName,
+                    existingNames = processedColumnNames,
+                    originalColumnName = columnName
+                )
+
+            processedColumnNames.add(finalColumnName)
+            columnMappings[columnName] = finalColumnName
```

Comment:
this doesn't seem right - columnMappings should be populated with the displayName, not the canonical name

I think the easiest diff would be:
* `resolveColumnNameCollision` takes a full ColumnName object instead of just the canonical name
* `resolveColumnNameCollision` _returns_ a full ColumnName object
* similar changes for superResolve